src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -3124,90 +3055,221 @@ namespace stan {
+                                         bool& pass,
+                                         std::ostream& error_msgs) const {
+      if (var_decl.bare_type().is_ill_formed_type()) {
+        error_msgs << "Variable declaration is ill formed,"
+                   << " variable name="
+                   << var_decl.name()
+                   << std::endl;
+        pass = false;
+        return;
+      }
+    }
+    boost::phoenix::function<validate_single_local_var_decl>
+    validate_single_local_var_decl_f;
+
+    void validate_array_local_var_decl::operator()(
+                                        local_var_decl& var_decl_result,
+                                        const local_var_type& el_type,
+                                        const std::string& name,
+                                        const std::vector<expression>& dims,
+                                        const expression& def,
+                                        bool& pass,
+                                        std::ostream& error_msgs) const {
+      if (dims.size() == 0) {
+        error_msgs << "Array type requires at least 1 dimension,"
+                   << " none found" << std::endl;
+        pass = false;
+        return;
+      }
+      if (el_type.bare_type().is_ill_formed_type()) {
+        error_msgs << "Array variable declaration is ill formed,"
+                   << " variable name="
+                   << name
+                   << std::endl;
+        pass = false;
+        return;
+      }
+      stan::lang::local_array_type bat(el_type, dims);
+      local_var_decl result(name, bat, def);
+      var_decl_result = result;
+    }
+    boost::phoenix::function<validate_array_local_var_decl>
+    validate_array_local_var_decl_f;
+
+    void validate_fun_arg_var::operator()(var_decl& var_decl_result,
+                                      const bare_expr_type& bare_type,
+                                      const std::string& name,
+                                      bool& pass,
+                                      std::ostream& error_msgs) const {
+      if (bare_type.is_ill_formed_type()) {
+        error_msgs << "Function argument is ill formed,"
+                   << " name="
+                   << name
+                   << std::endl;
+        pass = false;
+        return;
+      }
+      stan::lang::var_decl vd(name, bare_type);
+      var_decl_result = vd;
+    }
+    boost::phoenix::function<validate_fun_arg_var>
+    validate_fun_arg_var_f;
+
+    void validate_bare_type_is_data::operator()(
+                                       bare_expr_type& bare_type,
+                                       scope& scope,
+                                       bool& pass,
+                                       std::ostream& error_msgs) const {
+      if (bare_type.is_int_type()) {
+          pass = false;
+          error_msgs
+            << "Data qualifier cannot be applied to variable of type int"
```

Comment:
[future]
Maybe we should rethink this.  There are ways to have data change based on parameters, and we should have data qualifiers for integers going forward.  Would you mind opening an issue for that, too?  Thanks.

Reply:
checked code - this isn't called from anywhere - this compiles and runs:
```
functions {
  real foo(data int a) {
    return 0.0;
  }
}
```

Reply:
removed function from semantic actions altogether.